### PR TITLE
include a generated externalregistryhostname field

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@deads2k ptal. I think we want this so the registry operator can generate/populate it w/ the default route, and if the admin decides to expose the registry via some other route, they can set the spec field(already exists).

File: config/v1/types_image.go [hunk 1]
```diff
@@ -46,6 +46,12 @@ type ImageStatus struct {
 	// For backward compatibility, users can still use OPENSHIFT_DEFAULT_REGISTRY
 	// environment variable but this setting overrides the environment variable.
 	InternalRegistryHostname string `json:"internalRegistryHostname,omitempty"`
+
+	// ExternalRegistryHostname sets the hostname for the default external image
+	// registry. The external hostname should be set only when the image registry
+	// is exposed externally. The value is used in 'publicDockerImageRepository'
+	// field in ImageStreams. The value must be in "hostname[:port]" format.
+	ExternalRegistryHostname string `json:"externalRegistryHostname,omitempty"`
```
Comment (deads2k) [inline config/v1/types_image.go]: I would expect this to be a slice.  Why would we constrain ourselves to a single external representation?

Comment (bparees) [reply config/v1/types_image.go]: only one value can be consumed by the product (see the publicDockerImageRepository field in imagestreams), providing a slice means:

1) consumers(such as imagestreams) have to decide which one to use and may pick different ones
2) implying multiple values have meaning when in reality only one does

If we want the ability to expose additional values (ie the admin has setup 15 more domains for their registry) I can see us adding a second field like `ExternalRegistryHostnameAliases []string`

But for the canonical value I think exactly one is appropriate (and it's what we allowed for prior to 4.0 w/o any issue that i'm aware of)

Comment (bparees) [reply config/v1/types_image.go]: @deads2k updated this to be a slice per our conversation, ptal.

**Comment by bparees** [comment]
(since this will block other people bumping origin i'm going to get the client-go and origin PRs lined up before you merge this).

File: config/v1/types_image.go [hunk 2]
```diff
@@ -46,6 +46,12 @@ type ImageStatus struct {
 	// For backward compatibility, users can still use OPENSHIFT_DEFAULT_REGISTRY
 	// environment variable but this setting overrides the environment variable.
 	InternalRegistryHostname string `json:"internalRegistryHostname,omitempty"`
+
+	// ExternalRegistryHostname sets the hostname for the default external image
+	// registry. The external hostname should be set only when the image registry
+	// is exposed externally. The value is used in 'publicDockerImageRepository'
+	// field in ImageStreams. The value must be in "hostname[:port]" format.
+	ExternalRegistryHostname []string `json:"externalRegistryHostname,omitempty"`
```
Comment (bparees) [inline config/v1/types_image.go]: @deads2k is there a rule on pluralizing this since it's an array now?

File: config/v1/types_image.go [hunk 3]
```diff
@@ -46,6 +46,12 @@ type ImageStatus struct {
 	// For backward compatibility, users can still use OPENSHIFT_DEFAULT_REGISTRY
 	// environment variable but this setting overrides the environment variable.
 	InternalRegistryHostname string `json:"internalRegistryHostname,omitempty"`
+
+	// ExternalRegistryHostname sets the hostname for the default external image
```
Comment (deads2k) [inline config/v1/types_image.go]: api docs use the tag names `externalRegistryHostnames`

File: config/v1/types_image.go [hunk 4]
```diff
@@ -46,6 +46,12 @@ type ImageStatus struct {
...
+	// registry. The external hostname should be set only when the image registry
```
Comment (deads2k) [inline config/v1/types_image.go]: this doc is about the first one, right?

File: kubecontrolplane/v1/types.go [hunk 5]
```diff
@@ -80,7 +80,7 @@ type KubeAPIServerImagePolicyConfig struct {
 	// registry. The external hostname should be set only when the image registry
 	// is exposed externally. The value is used in 'publicDockerImageRepository'
 	// field in ImageStreams. The value must be in "hostname[:port]" format.
-	ExternalRegistryHostname string `json:"externalRegistryHostname"`
+	ExternalRegistryHostnames []string `json:"externalRegistryHostnames"`
```
Comment (deads2k) [inline kubecontrolplane/v1/types.go]: you can make this change at the same time if you like, but it means you need an origin PR green and read to go to handle it.

File: legacyconfig/v1/types.go [hunk 6]
```diff
@@ -413,7 +413,7 @@ type ImagePolicyConfig struct {
 	// registry. The external hostname should be set only when the image registry
 	// is exposed externally. The value is used in 'publicDockerImageRepository'
 	// field in ImageStreams. The value must be in "hostname[:port]" format.
-	ExternalRegistryHostname string `json:"externalRegistryHostname,omitempty"`
+	ExternalRegistryHostnames []string `json:"externalRegistryHostnames,omitempty"`
```
Comment (deads2k) [inline legacyconfig/v1/types.go]: this one should not change.  This format is used to read existing masterconfigs.

File: openshiftcontrolplane/v1/types.go [hunk 7]
```diff
@@ -83,7 +83,7 @@ type ImagePolicyConfig struct {
 	// registry. The external hostname should be set only when the image registry
 	// is exposed externally. The value is used in 'publicDockerImageRepository'
 	// field in ImageStreams. The value must be in "hostname[:port]" format.
-	ExternalRegistryHostname string `json:"externalRegistryHostname"`
+	ExternalRegistryHostnames []string `json:"externalRegistryHostnames"`
```
Comment (deads2k) [inline openshiftcontrolplane/v1/types.go]: same comment about origin

**Comment by deads2k** [comment]
other than doc, the config change looks good.  The legacyconfig isn't allowed to change.  It must match 3.11 forever.  the kubecontrolplane and openshfitcontrolplane must have a paired origin pull that is green and read to merge. 

You can split the config change out ahead of the rest if you like.

File: kubecontrolplane/v1/types.go (hunk 5 above)
Comment (bparees) [reply kubecontrolplane/v1/types.go]: yeah i'm working on the origin PR.

**Comment by bparees** [comment]
@deads2k updated.  Please sign off, I realize it can't merge until origin is updated but i've already wasted a ton of time iterating on origin changes that have to be redone every time this PR changes.

**Comment by bparees** [comment]
github notifications are slow. thanks @deads2k i've pushed the changes you requested.

**Comment by bparees** [comment]
@openshift/api-reviewers ptal.  the origin PR is here and passing tests:

https://github.com/openshift/origin/pull/21490

so after this is merged i will bump origin to pick up this change.

**Comment by bparees** [comment]
@openshift/api-reviewers bump.  This has a matching origin PR that is passing tests:
https://github.com/openshift/origin/pull/21490

so as soon as it is merged, I can go bump origin properly.

**Comment by sjenning** [comment]
@bparees a never found a reason for the change.  why are we switching to allow multiple external registries?

**Comment by sjenning** [comment]
nm, this feeds the image pull policy admission plugin.  i see now.

i'll let @deads2k approve since this is his component's config... and i'm a chicken :hatching_chick:

**Pull request merged**